src/lm/participants.ts

```diff
@@ -0,0 +1,196 @@
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+import * as vscode from 'vscode';
+import { dispose } from '../common/utils';
+import { IToolCall } from './tools/toolsUtils';
+
+const llmInstructions = `Instructions:
+- The user will ask a question related to GitHub, and it may require lots of research to answer correctly. There is a selection of tools that let you perform actions or retrieve helpful context to answer the user's question.
+- If you aren't sure which tool is relevant, you can call multiple tools. You can call tools repeatedly to take actions or gather as much context as needed until you have completed the task fully. Don't give up unless you are sure the request cannot be fulfilled with the tools you have.
+- Don't ask the user for confirmation to use tools, just use them.
+- When talking about issues, be as concise as possible while still conveying all the information you need to. Avoid mentioning the following:
+  - The fact that there are no comments.
+  - Any info that seems like template info.`;
+
+export class ChatParticipantState {
+	private _messages: vscode.LanguageModelChatMessage[] = [];
+
+	get lastToolResult(): (string | vscode.LanguageModelToolResultPart | vscode.LanguageModelToolCallPart)[] {
+		for (let i = this._messages.length - 1; i >= 0; i--) {
+			const message = this._messages[i];
+			for (const part of message.content2) {
+				if (part instanceof vscode.LanguageModelToolResultPart) {
+					return message.content2;
+				}
+			}
+		}
+		return [];
+	}
+
+	get firstUserMessage(): string | undefined {
+		for (let i = 0; i < this._messages.length; i++) {
+			const message = this._messages[i];
+			if (message.role === vscode.LanguageModelChatMessageRole.User && message.content) {
+				return message.content;
+			}
+		}
+	}
+
+	get messages(): vscode.LanguageModelChatMessage[] {
+		return this._messages;
+	}
+
+	addMessage(message: vscode.LanguageModelChatMessage): void {
+		this._messages.push(message);
+	}
+
+	reset(): void {
+		this._messages = [];
+	}
+
+	constructor() {
+
+	}
+}
+
+export class ChatParticipant implements vscode.Disposable {
+	private readonly disposables: vscode.Disposable[] = [];
+
+	constructor(context: vscode.ExtensionContext, private readonly state: ChatParticipantState) {
+		const ghprChatParticipant = vscode.chat.createChatParticipant('githubpr', (
+			request: vscode.ChatRequest,
+			context: vscode.ChatContext,
+			stream: vscode.ChatResponseStream,
+			token: vscode.CancellationToken
+		) => this.handleParticipantRequest(request, context, stream, token));
+		ghprChatParticipant.iconPath = vscode.Uri.joinPath(context.extensionUri, 'resources/icons/github_logo.png');
+		this.disposables.push(ghprChatParticipant);
+	}
+
+	dispose() {
+		dispose(this.disposables);
+	}
```

Comment:
nit: Is there a base class to extends instead of this?

Reply:
Nope, but we need one. Will tackle as debt across the codebase: https://github.com/microsoft/vscode-pull-request-github/issues/6285